step 1 - Get annotations for all the species 
generate_annotations.pl -sp sp1 -GTF sp1.gtf -G sp1.fasta 
generate_annotations.pl -sp sp2 -GTF sp2.gtf -G sp2.fasta

step 2 - Split cluster file by pair of species

perl get_gcl_sp_pair.pl gene_cluster_file.txt

step 3 - split cluster of species pair in parts according to the number of alignments and submit the alignments.

perl Prepare_and_Submit_Aln_sp_pair_V2.pl -s1 sp1 -s2 sp2 --expath path_to_annotations --gene_cluster gene_cluster_file.txt --N_split number_of_alignments_split_files --project_dir output_directory --bin bin_directory

=example=

perl Prepare_and_Submit_Aln_sp_pair_V2.pl -s1 Ame -s2 Cdi --gene_cluster GCL_Ame-Cdi.tab --project_dir ./ --bin exon_intron_orthology_pipeline/bin -expath /users/mirimia/ymarquez/EXONS_DB --N_split 1000

step 4

Realign_target_exons_by_part.pl sp1 sp2 exons_to_realign_part_$i.txt exint_file_sp1 exint_file_sp2 $i outfile output_folder_sp1_sp2 blosum62_file bin_folder

outfile has to be named: realigned_exons_sp1_sp2_<$i>.txt

e.g. 
Parameters:

- sp1  Ame 
- sp2  Cdi 
- infile exons_to_realign_part_1.txt (output generated by Prepare_and_Submit_Aln_sp_pair.pl, one part for each cluster part)
- exint_file_sp1 Ame.exint
- exint_file_sp2 Cdi.exint
- part 1 (several parts, each for each cluster part)
- outfile  realigned_exons_Ame_Cdi_1.txt (has to have this name and include the part of the cluster also in the name file)
- output_folder  Ame_Cdi
- blosum62_file  files/blosum62.txt
- bin_folder bin/

Realign_target_exons_by_part.pl Ame Cdi exons_to_realign_part_1.txt Ame.exint Cdi.exint 1 realigned_exons_Ame_Cdi_1.txt Ame_Cdi files/blosum62.txt bin/


step 5
collecting evertything
output cluster of exons.

